### ISSUE TITLE
Error handling when creating instances via a builder

### ISSUE THREAD

**Issue by WhyNotHugo** [z-galaxy/zbus#403]
I'm bringing this up as a response to the toot:

> If you want changes that require API breakage, now would be a good time to speak out

Something that has low-key bothered me was error handling in builders. In particular, I often have operations that I know are infallible but I need to handle errors due to their API (generally, I just add `.expect("this cannot fail")`, but that's usually an indicator of an API not being quite right.

For example: https://docs.rs/zbus/latest/zbus/struct.ProxyBuilder.html#method.path

If I already have a Path instance, I know that calling this method is infallible, but I still need to handle (ignore?) the error.

### First potential approach

The first one is the more-API-breaking one, and looks something like:

```rust
pub fn path(self, path: ObjectPath) -> Self;

pub fn try_path<P>(self, path: P) -> Result<Self>where
    P: TryInto<ObjectPath<'a>>,
    P::Error: Into<Error>;
```

The major downside of this one is that it requires a lot of work for applications to adapt. 

### Second potential approach

The second approach is less invasive, and to just update the API to:

```rust
pub fn path<P>(self, path: P) -> Self where
    P: TryInto<ObjectPath<'a>>,
```

And then have `ProxyBuilder.build` return an error if the Path was invalid. This approach might get trickier with error types.

### Third potential approach

The third approach is likely the weirdest, but least invasive, and most compatible, by introducing a new function:

```rust
pub fn from_parts(self, path: Path, interface: InterfaceName, destination: BusName) -> T;
```

Where all the "parts" (for lack of a better name) have already been validated, this function cannot fail, and matches what's going on.

**Comment by zeenix** [comment]
Hi @WhyNotHugo,

Thanks for reporting this and also for suggesting solutions.

Given that the main idea behind keeping these generic is convenience, solution 2 seems most interesting at first thought.

> This approach might get trickier with error types.

I guess we can keep the errors generic to (like they're right now)? This will mean us having to keep everything in a Box though but I think the cost of allocation is still too small here to worry about at all and the gain is quite big.

This will still be a breaking change though but super easy to fix. What I'm concerned about though, is that for many it will require  a lot of work (even if it's very trivial changes).  Even though we're breaking API, I'd really prefer to not break most of existing code, at least not in a way that requires a lot of work to port.

So maybe we should consider:

1. The option 1 above but inverted and with different naming. i-e keep the existing methods but introduce new alternatives that take a ready object. This is likely the best option, given we can come up with nice-enough names. I can't think of any right now.
2. The option 3 but we need the method being fallible (since the point is to avoid having to `unwrap` for each part) so we don't have to Box anything and the new method works through existing machinery.

**Comment by WhyNotHugo** [comment]
> The option 1 above but inverted and with different naming. i-e keep the existing methods but introduce new alternatives that take a ready object. This is likely the best option, given we can come up with nice-enough names. I can't think of any right now.

I thought about this, but it kinda goes against the usual conventions where `from` is infallible and `try_from` is fallible (same for `try_clone` and `clone`), etc.

> The option 3 but we need the method being fallible (since the point is to avoid having to unwrap for each part) so we don't have to Box anything and the new method works through existing machinery.

Why would this method be fallible? In what situation can it fail?

**Comment by zeenix** [comment]
> > The option 1 above but inverted and with different naming. i-e keep the existing methods but introduce new alternatives that take a ready object. This is likely the best option, given we can come up with nice-enough names. I can't think of any right now.
> 
> I thought about this, but it kinda goes against the usual conventions where `from` is infallible and `try_from` is fallible (same for `try_clone` and `clone`), etc.

Yeah, I know. It's just that it will require absolutely no change in client code.

> > The option 3 but we need the method being fallible (since the point is to avoid having to unwrap for each part) so we don't have to Box anything and the new method works through existing machinery.
> 
> Why would this method be fallible? In what situation can it fail?

You're right. I had missed that this method is taking ready objects.

**Comment by zeenix** [comment]
@WhyNotHugo looking into this now. I'm leaning towards the first option. Still can't think of a good name though. Can you?

**Comment by zeenix** [comment]
@skullone suggested `raw_` prefix on the Matrix channel. I think I like that.

**Comment by zeenix** [comment]
ok, since I'll be going with the non-breaking option, I'm lowering the priority here. Also this could be a good/easy way for anyone to contribute.

**Comment by WhyNotHugo** [comment]
“raw” is a weird naming for interfaces that take a built/parsed input. “raw” makes me think that it takes the exact opposite: a non-parsed / non-built (i.e.: raw) input.

The typical Rust convention is to call fallible methods try_something() and the infallible one just something().

**Comment by zeenix** [comment]
> The typical Rust convention is to call fallible methods try_something() and the infallible one just something().

We already discussed this. That breaks every single code out there and I really want to avoid that. Let me paste the relevant discussion from the Matrix:

```
zeenix: I need advice on naming. in the ProxyBuilder, we've generic setters for each attribute, path, interface and destination and they're all currently fallible. I want to provide non-fallible versions that take a ready instance of the type (e.g ObjectPath instead of O: TryInto<ObjectPath>).
if it didn't break all existing code, I'd prefer to rename the current ones to prefix them with try_
but we need to do the opposite for breaking existing code completely

* Bilal Elmoussaoui (@bilelmoussaoui:gnome.org) opts for breaking existing code

zeenix: there is also another issue: people passing strings will now have to write try_ for each method call
I think they're a majority (educated guess)

skullone:  There's precedent for a raw_ prefix in https://doc.rust-lang.org/std/os/windows/process/trait.CommandExt.html#tymethod.raw_arg for raw_arg vs arg of the former not doing processing that the latter does.
[CommandExt in std::os::windows::process - Rust](https://doc.rust-lang.org/std/os/windows/process/trait.CommandExt.html#tymethod.raw_arg)
Windows-specific extensions to the `process::Command` builder.
if it weren't just setting a value I'd semi-seriously suggest a do_ prefix in comparison to a try_ prefix, but mostly just to attribute the suggestion "do, or do-not, there is no try" to Yoda
```

**Comment by zeenix** [comment]
> “raw” is a weird naming for interfaces that take a built/parsed input. “raw” makes me think that it takes the exact opposite: a non-parsed / non-built (i.e.: raw) input.

Maybe that's not the best name, yeah but so please do propose alternative naming.

**Comment by MaxVerevkin** [comment]
I did some `rg`ing to have statistics on how `i3status-rust` is using these APIs. I got

### Path

Type | Count
---|---
Literal | 3
String | 5
ObjectPath | 6

### Destination

Type | Count
---|---
Literal | 3
BusName | 1

### Interface

Type | Count
---|---
Literal | 5

It is worth noting that some of these literals are duplicates (for example `org.freedesktop.DBus.Properties` is used 4 times), and they would better off be (typed) constants.

Regarding typed constants, they would require `const fn` constructors, but I believe it is already possible in stable rust, with some limitations (for example instead of using `Error::InvalidInterfaceName(String)` of all errors, a new enum listing all possible errors would be needed, which arguably is a good thing).

Passing string literals seem to be one of the major reasons why the current API is what it is, and changing it to `pub fn path(self, path: ObjectPath) -> Self;` will make using literals not as ergonomic. This is completely true. But maybe it is a good thing? Together with typed consts, it would encourage people (myself included) to better organize their code and catch mistakes at compile time (not to mention the removal of useless error handling).

**Comment by zeenix** [comment]
> Regarding typed constants, they would require `const fn` constructors, but I believe it is already possible in stable rust, with some limitations (for example instead of using `Error::InvalidInterfaceName(String)` of all errors, a new enum listing all possible errors would be needed, which arguably is a good thing).

I'm not sure I follow your exact proposal on typed constants. Can you elaborate with a client code example?

> Passing string literals seem to be one of the major reasons why the current API is what it is, and changing it to `pub fn path(self, path: ObjectPath) -> Self;` will make using literals not as ergonomic. This is completely true.

Correct. What makes things complicated is that I want to support both concrete type and a string. It doesn't necessarily have to be a literal though. There are many services out there that create the object path on the fly.

> But maybe it is a good thing? Together with typed consts, it would encourage people (myself included) to better organize their code and catch mistakes at compile time (not to mention the removal of useless error handling).

Maybe but I'll need to understand what you mean exactly by typed constants. :) Just FYI, there is already agreement that we need macros that create the concrete types at compile time hence catch any errors while user type them (e.g while writing macro code) (#316) and @WhyNotHugo even put a lot of work on this [a year ago](https://gitlab.freedesktop.org/dbus/zbus/-/merge_requests/486) but at the time, we weren't sure about crate organization so it didn't get anywhere.

**Comment by MaxVerevkin** [comment]
> I'm not sure I follow your exact proposal on typed constants

I mean it would be nice to be able to write

```rust
const DBUS_PROPS_IFACE: InterfaceName = InterfaceName::new("org.freedesktop.DBus.Properties");
```

instead of

```rust
const DBUS_PROPS_IFACE: &str = "org.freedesktop.DBus.Properties";
```

In this example the validation happens only once and at compile time.

**Comment by MaxVerevkin** [comment]
One more idea. Since we now have GATs, we can do

```rust
let path_name: PathName = todo!();

let fallible: Result<_, _> = Builder.path("string");
let infallible: Builder = Builder.path(path_name);
```

```rust
trait IntoPathName {
    type Result<T, E>;
    type Err;

    fn into_path_name(self) -> Self::Result<PathName, Self::Err>;
}

impl IntoPathName for &'_ str {
    type Result<T, E> = Result<T, E>;
    type Err = Error;

    fn into_path_name(self) -> Self::Result<PathName, Self::Err> {
        todo!()
    }
}

impl IntoPathName for PathName {
    type Result<T, E> = T;
    type Err = Infallible;

    fn into_path_name(self) -> Self::Result<PathName, Self::Err> {
        self
    }
}

impl Builder {
    fn path<P>(self, path: P) -> P::Result<Self, P::Err>
    where
        P: IntoPathName,
        P::Err: Into<Error>,
    {
        todo!()
    }
}
```

This is just a quick sketch of what the API might look like. I think it's the first real world usecase of GATs that don't involve lifetimes that I came across.

**Comment by zeenix** [comment]
> Edit: i got something working, but it is not as pretty as I hoped: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=61934865dabb61ec8cd3aced24106a59

Very interesting. Thanks for explaining. However, I also think this is a bit too complicated (if it was one type, I'd say let's bite the bullet and go for it but we've at least 8 types). Maybe we can think more about this and come up with something simpler.

**Comment by zeenix** [comment]
ok, given that:

* most users use macros anyway
* `TryInto` is now in the std prelude since Rust 2021.

How about we switch:

* most of low-level API to take concrete types.
* later (after 4.0 as this won't require a break) provide macros to construct the types at compile time and use them in `dbus_proxy` and `dbus_interface` macros to give users instant error as they type (#230). I hope @WhyNotHugo can help here.

Unfortunately, the `proxy::Builder` is an API that users use directly even with macros. So I'm re-thinking about option 1 here. I wonder if the impact is really as high as we're thinking. Since interface is mandatory in `dbus_proxy`, we only have destination and path left. I'm willing to bet destination is typically also static. When object path is not static, it typically comes from another method call (or property) and users can easily declare the return value of those methods/properties to be `ObjectPath` type. 

Given that and the fact that changing from `builder.path(p)" to `builder.try_path(p)` is trivial (and can be mostly automated), perhaps option 1 is best here?

**Comment by MaxVerevkin** [comment]
> However, I also think this is a bit too complicated

I agree.

> but we've at least 8 types

Here is my attempt to generalize this pattern by creating a `TryInto2` trait: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=faf6b552a1b552a156148c82f901b9a7

**Comment by MaxVerevkin** [comment]
What are your thoughts on `MatchRule::builder`?

**Comment by zeenix** [comment]
> What are your thoughts on `MatchRule::builder`?

I think that's not something most users use (at least they shouldn't need to) so applying option 1 above to that, is not a problem.

**Comment by zeenix** [comment]
> Here is my attempt to generalize this pattern by creating a `TryInto2` trait: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=faf6b552a1b552a156148c82f901b9a7

I sort of like it. It's going to be quite a bit of work for us though. I'm already running out of time. :(

**Comment by MaxVerevkin** [comment]
> It's going to be quite a bit of work for us though. I'm already running out of time. :(

Is there a deadline for v4.0?

**Comment by zeenix** [comment]
> > It's going to be quite a bit of work for us though. I'm already running out of time. :(
> 
> Is there a deadline for v4.0?

Not exactly but I would like to roll it out soon (1-2 weeks, if possible). What I mainly meant was time I have for working on zbus. :)

**Comment by zeenix** [comment]
> > What are your thoughts on `MatchRule::builder`?
> 
> I think that's not something most users use (at least they shouldn't need to) so applying option 1 above to that, is not a problem.

On second though, people who do use MatchRule would typically just use strings so perhaps we want the other way around in there: convert them to take `&st`.

**Comment by MaxVerevkin** [comment]
> perhaps we want the other way around in there: convert them to take &st.

What would be the benefit of this change? The return type will be `Result` anyway.

**Comment by zeenix** [comment]
> > perhaps we want the other way around in there: convert them to take &st.
> 
> What would be the benefit of this change? The return type will be `Result` anyway.

My thinking was that we'll reduce binary bloat that way but [after changing some of API like this](https://github.com/dbus2/zbus/commit/bbd441fc7fa93048e1f0e1d0fb241b255a7d40e5), I realized will not have the desired effects because if all calls are passing a `&str` then the method will be expanded into only 1 version of it.

**Comment by zeenix** [comment]
> > but we've at least 8 types
> 
> Here is my attempt to generalize this pattern by creating a `TryInto2` trait: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=faf6b552a1b552a156148c82f901b9a7

I made some progress on making it slightly simpler: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d3de185f764a901dd2a6fab410ce002e but I didn't get too far with avoiding the trait. Let's assume we want to do this, what would be the exact plan for our crates and most importantly, how do we name this trait? :laughing:

**Comment by MaxVerevkin** [comment]
> most importantly, how do we name this trait?

Good question:) IDK...

> Let's assume we want to do this

I'm still not quite sure about this. Mostly because GATs is a relatively new and unused feature, so it may be confusing to many. I would like more people to confirm that this is a sane idea.

Also specialization would be really nice here. We could have a default implementation for `T: TryInto<U>` and a specialized `T: Into<U>`. Really sad that specialization is nowhere near stable.

**Comment by zeenix** [comment]
> > Let's assume we want to do this
> 
> I'm still not quite sure about this. Mostly because GATs is a relatively new and unused feature, so it may be confusing to many. I would like more people to confirm that this is a sane idea.

From the user's POV, I don't think they need to really understand how it all works. We can always do a good job of documentation and it should be fine for the typical user. Especially examples in the builder docs would really help. They just need to know what types they can pass and when the method is infallible.

> Also specialization would be really nice here. We could have a default implementation for `T: TryInto<U>` and a specialized `T: Into<U>`. Really sad that specialization is nowhere near stable.

That would be nice indeed. Given all of that, the fact that I don't think having to type "?" after each builder method even in case they can't fail, is not that big a deal and switching from generics to concrete types doesn't help with binary size, I will have to give this a miss for 4.0.

Let's leave it for the next API break. :)
